File: Fundamentals_of_Software_Design/2_Algorithms/Поиск подстроки в строке алгоритмом Кнута-Морриса-Пратта/main.py
def compute_lps(needle):
    """Вычисляет массив LPS (longest prefix suffix) для подстроки needle."""
    lps = [0] * len(needle)
    length = 0
    i = 1

    while i < len(needle):
        if needle[i] == needle[length]:
            length += 1
            lps[i] = length
            i += 1
        elif length:
            length = lps[length - 1]
        else:
            lps[i] = 0
            i += 1
    return lps

def my_strstr(haystack, needle):
    """Алгоритм Кнута-Морриса-Пратта (КМП) для поиска подстроки."""
    if not needle:
        return 0

    n, m = len(haystack), len(needle)
    lps = compute_lps(needle)
    i = j = 0  # Индексы для haystack и needle

    while i < n:
        if haystack[i] == needle[j]:
            i += 1
            j += 1
        if j == m:
            return i - j  # Найдено вхождение
        elif i < n and haystack[i] != needle[j]:
            j = lps[j - 1] if j else 0
            i += not j  # Увеличиваем i, если j == 0
    return None  # Не найдено

# --- Тесты ---
def test_my_strstr():
    assert my_strstr("hello", "ll") == 2
    assert my_strstr("abcdef", "cd") == 2
    assert my_strstr("aaaaa", "bba") is None
    assert my_strstr("mississippi", "issi") == 1
    assert my_strstr("abc", "") == 0
    assert my_strstr("", "abc") is None
    assert my_strstr("abc", "abc") == 0
    print("Все тесты пройдены!")

test_my_strstr()

File: Fundamentals_of_Software_Design/2_Algorithms/Поиск подстроки в строке алгоритмом Кнута-Морриса-Пратта/main.c
#include <stdio.h>
#include <string.h>

void compute_lps(const char *needle, int m, int *lps) {
    int length = 0, i = 1;
    lps[0] = 0;

    while (i < m) {
        if (needle[i] == needle[length]) {
            lps[i++] = ++length;
        } else if (length) {
            length = lps[length - 1];
        } else {
            lps[i++] = 0;
        }
    }
}

char *my_strstr(char *haystack, const char *needle) {
    if (!*needle) return haystack; // Если needle пустая, вернуть haystack

    int n = strlen(haystack), m = strlen(needle);
    if (m > n) return NULL;

    int lps[m];
    compute_lps(needle, m, lps);

    int i = 0, j = 0;
    while (i < n) {
        if (haystack[i] == needle[j]) {
            i++, j++;
            if (j == m) return haystack + i - j;
        } else {
            j = j ? lps[j - 1] : 0;
            if (!j) i++;
        }
    }
    return NULL;
}

// --- Тесты ---
void run_tests() {
    struct { char *haystack; char *needle; int expected; } tests[] = {
        {"hello", "ll", 2},
        {"abcdef", "cd", 2},
        {"aaaaa", "bba", -1},
        {"mississippi", "issi", 1},
        {"abc", "", 0},
        {"", "abc", -1},
        {"abc", "abc", 0}
    };
    
    int num_tests = sizeof(tests) / sizeof(tests[0]);
    for (int i = 0; i < num_tests; i++) {
        char *result = my_strstr(tests[i].haystack, tests[i].needle);
        int pos = result ? result - tests[i].haystack : -1;
        printf("Test %d: %s\n", i + 1, pos == tests[i].expected ? "+" : "-");
    }
}

int main() {
    run_tests();
    return 0;
}
